stage0/src/pic.rs

```diff
@@ -0,0 +1,78 @@
+//
+// Copyright 2023 The Project Oak Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+
+use oak_sev_guest::io::{IoPortFactory, PortFactoryWrapper, PortWrapper, PortWriter};
+
+const PIC0_BASE: u16 = 0x20;
+const PIC1_BASE: u16 = 0xA0;
+
+pub struct Pic {
+    command: PortWrapper<u8>,
+    data: PortWrapper<u8>,
+}
+
+impl Pic {
+    pub fn new(port_factory: &PortFactoryWrapper, base: u16) -> Self {
+        Self {
+            command: port_factory.new_writer(base),
+            data: port_factory.new_writer(base + 1),
+        }
+    }
+
+    pub unsafe fn write_command(&mut self, command: u8) -> Result<(), &'static str> {
+        self.command.try_write(command)
+    }
+
+    pub unsafe fn write_data(&mut self, data: u8) -> Result<(), &'static str> {
+        self.data.try_write(data)
+    }
+
+    pub unsafe fn init(
+        &mut self,
+        interrupt_offset: u8,
+        chaining: u8,
+        mask: u8,
+    ) -> Result<(), &'static str> {
+        // The initialization process is documented in https://wiki.osdev.org/8259_PIC
+        // ICW1: ICW_INIT | ICW_ICW4
+        self.write_command(0x11)?;
+        // ICW2: the interrupt offset.
+        self.write_data(interrupt_offset)?;
+        // ICW3, "master/slave device": chaining between PICs
```

Comment:
nit: prefer "leader/follower device" or "initiator/responder device"

Reply:
This is straight from the datasheet printed in the 70s, it was a different time :)